src/analytics/src/main/scala/osmesa/analytics/stats/CalculateStats.scala

```diff
@@ -0,0 +1,486 @@
+        select(
+          $"id",
+          $"changeset",
+          $"version",
+          $"members",
+          statTopics($"tags", "relation").as("statTopics")
+        )
+
+    val relationStatChanges: RDD[(Long, StatCounter)] =
+      relationInfo.
+        where(size($"statTopics") > 0).
+        select($"id", $"changeset", $"version", $"statTopics").
+        rdd.
+        map { row =>
+          val osmId = RelationId(row.getAs[Long]("id"))
+          val changeset = row.getAs[Long]("changeset")
+          val version = row.getAs[Long]("version")
+          val topics = row.getAs[Seq[StatTopic]]("statTopics").toSet
+          (changeset, StatCounter(osmId, changeset, version, topics))
+        }.
+        partitionBy(wayPartitioner)
+
+    val relationMembers =
+      relationInfo.
+        select($"id", $"changeset", $"version", $"statTopics", explode($"members").as("member"))
+
+    val relationsForWays: RDD[(Long, List[(RelationId, UpstreamInfoMap)])] =
+      generateUpstreamInfoMaps {
+        relationMembers.
+          where($"member.type" === "way").
+          select($"id", $"changeset", $"version", $"statTopics", $"member.ref".as("wayId")).
+          rdd.
+          map { row =>
+            val osmId = RelationId(row.getAs[Long]("id"))
+            val changeset = row.getAs[Long]("changeset")
+            val version = row.getAs[Long]("version")
+            val topics = row.getAs[Seq[StatTopic]]("statTopics").toSet
+            val wayId = row.getAs[Long]("wayId")
+
+            (wayId, (osmId, changeset, version, topics))
+          }.
+          groupByKey(wayPartitioner)
+      }
+
+    val relationsForNodes: RDD[(Long, List[(RelationId, UpstreamInfoMap)])] =
+      generateUpstreamInfoMaps {
+        relationMembers.
+          where($"member.type" === "node").
+          select($"id", $"changeset", $"version", $"statTopics", $"member.ref".as("nodeId")).
+          rdd.
+          map { row =>
+            val osmId = RelationId(row.getAs[Long]("id"))
+            val changeset = row.getAs[Long]("changeset")
+            val version = row.getAs[Long]("version")
+            val topics = row.getAs[Seq[StatTopic]]("statTopics").toSet
+            val nodeId = row.getAs[Long]("nodeId")
+
+            (nodeId, (osmId, changeset, version, topics))
+          }.
+          groupByKey(nodePartitioner)
+      }
+
+    val wayInfo =
+      history.
+        where("type == 'way'").
+        select(
+          $"id",
+          $"changeset",
+          $"version",
+          $"timestamp",
+          $"nds.ref".as("nodes"),
+          statTopics($"tags", "way").as("statTopics")
+        ).
+        rdd.
+        map { row =>
+          val id = row.getAs[Long]("id")
+          val changeset = row.getAs[Long]("changeset")
+          val version = row.getAs[Long]("version")
+          val instant = row.getAs[Timestamp]("timestamp").toInstant.toEpochMilli
+          val nodeIds = row.getAs[Seq[Long]]("nodes").toArray
```

Comment:
@lossyrob this is one of 2 spots that I suspect may be blowing up Kryo (when there are lots of nodes associated with a way). This is bypassed in the relation handling by using `explode` to create individual rows.

Reply:
The largest way has 49,189 refs.

(The most tags a single element has is 7,838. The most members a relation has is 69,643.)